### PR TITLE
[setup] migrate setup script to `pyproject.toml`

### PR DESCRIPTION
**Pull request by XuehaiPan**
# What does this PR do?

Follows discussion in https://github.com/huggingface/transformers/pull/22531#issuecomment-1494493545

Changes:

- migrate setup script to `pyproject.toml`
- migrate `pytest` configs to `pyproject.toml`
- cleanup `isort` and `flake8` configs

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -426,36 +426,7 @@ def run(self):
 setup(
     name="transformers",
     version="4.28.0.dev0",  # expected format is one of x.y.z.dev0, or x.y.z.rc1 or x.y.z (no to dashes, yes to dots)
-    author="The Hugging Face team (past and future) with the help of all our contributors (https://github.com/huggingface/transformers/graphs/contributors)",
-    author_email="transformers@huggingface.co",
-    description="State-of-the-art Machine Learning for JAX, PyTorch and TensorFlow",
-    long_description=open("README.md", "r", encoding="utf-8").read(),
-    long_description_content_type="text/markdown",
```
Comment (sgugger) [inline setup.py:433]: We are losing those apparently.

**Review by sgugger — COMMENTED** [review]
Thanks for your PR. Will double check that the package data keeps to be included. Other than that flagged one difference.

File: setup.py (hunk 1 above)
Comment (XuehaiPan) [reply setup.py:433]: It goes to `project.readme` in `pyproject.toml`. The content type will be inferred from the extension (`.md` or `.rst`).

**Review by sgugger — APPROVED** [review]
Tested locally and can confirm all extra files are still included in the build. Thanks for your PR!

**Pull request merged**